### ISSUE TITLE
Error with torch.dtype=float16

### ISSUE THREAD

**Issue by Patrick-Ni** [rachtibat/LRP-eXplains-Transformers#7]
Hello,
When i try this code: https://lxt.readthedocs.io/en/latest/quickstart.html#tinyllama in my gpu server, i found that an error:
<img width="1292" alt="image" src="https://github.com/user-attachments/assets/1edea5ee-d10a-461b-98e0-ef5d480ae414">

However, when i change the torch_dtype in ``model = LlamaForCausalLM.from_pretrained("llama-2-7b-chat-hf", torch_dtype=torch.float16, device_map='auto', quantization_config=quantization_config)`` to bfloat16, it works. 
I wonder that if i must use bf16? or there exists a way to fix this "bug"?

**Comment by Patrick-Ni** [comment]
And by the way, how can i get lrp score for each layer in llama?

**Comment by rachtibat** [comment]
Hey @Patrick-Ni,

since we are computing gradients, we need the bfloat16 datatype which has a bigger numerical range. (The models are also trained in bfloat16, which prevents numerical overflow etc.).
So, we cant use float16, we must use bfloat16.

Yes, you can simply apply a backward hook on the LRP rules or modules, which are normal torch modules.
E.g. if you print the Llama model after you registered the attnlrp rules, you see on which modules you can apply backward hooks. In Llama we have lxt.modules.LinearEpsilon or lxt.modules.SoftmaxDT Modules on which you could apply hooks.

Btw, I would start using the standard Llama model without quantization for testing LRP.
If you want to use quantized weights, you need a different attnlrp Composite (like the one in the Mixtral example, but I can help you out another day regarding this).

Hope it helps,
Reduan

**Comment by Patrick-Ni** [comment]
This has resolved most of my issues. Thank you for your helpful response!
